### PR TITLE
[browser] fix OGR provider not used to scan archives

### PR DESCRIPTION
**Pull request by nirvn**
## Description
After updating my workstation to upcoming ubuntu 18.10 (with gdal 2.3), my browser panel suddenly linked all archived files to the GDAL provider (when the archived file's scan behavior is set to basic). I'm not sure why this didn't happen under ubuntu 18.04 (looking into GDAL's release note, compressed raster reading support was already present in 18.04's gdal 2.2).

Anyhow, this PR insures that - when the archived file's scan mode is basic, i.e. archive actually not opened to insure compatibility with GDAL and/or OGR - we _always_ favor OGR (vector) over GDAL (raster) in our browser.

The alternative approach would have been to always open the archived file via GDALOpen / GDALOpenEx to test for raster support, but IMHO the potential CPU cost is too high, and reward too low. Enforcing a full archive inspection via those functions remains a possibility by setting the archive scan mode to full scan.

@nyalldawson , all good?

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by nirvn** [comment]
@elpaso , your thoughts on this would be appreciated too :)

**Comment by elpaso** [comment]
We'd miss the rasters from gpkg and spatialite, correct? I feel there will be complaints about not seeing those rasters in the browser... 
What about adding an option?

**Comment by nirvn** [comment]
@elpaso , OGR already prioritized for gpkg datasets already (https://github.com/qgis/QGIS/blob/master/src/providers/gdal/qgsgdaldataitems.cpp#L273)

**Comment by nirvn** [comment]
That said, I'm wondering whether there's an underlying issue with the browser here: shouldn't two providers provide two separate items in the browser?? I.e., if the archive method is basic, shouldn't there be two items in the browser panel for a given ZIP, i.e. GDAL and OGR items?

**Comment by elpaso** [comment]
> That said, I'm wondering whether there's an underlying issue with the browser here: shouldn't two providers provide two separate items in the browser?? I.e., if the archive method is basic, shouldn't there be two items in the browser panel for a given ZIP, i.e. GDAL and OGR items?

I think that if a source contains both rasters and vectors we should list them both and use `GDALOpenEx`.

But there is another open question: I think that some time ago, there was a discussion about a long-term goal to handle as much as possible through OGR/GDAL instead of using custom providers, the focus was spatialite, that's why the browser shows spatialite layers through OGR (which is a source of confusion btw).

**Comment by nirvn** [comment]
Well turns out the qgszipitem class was semi-broken, unable to register the OGR provider since the class was using the old data item creation function (which gdal still implemented until now, whereas ogr moved onto the new method).

This fixes a pretty significant issue, whereas if a archive had multiple vector datasets (or mixed vector / raster), the vector layers were not visible.

**Comment by nirvn** [comment]
Also, now that the code is fixed, the behavior has adjusted to the following: if a zip file has only one dataset, it'll defer to individual provider(s) to create the item, resulting in a zip file showing both as raster and vector (be @nyalldawson request), _unless_ the archive method is set to full scan.

**Comment by nirvn** [comment]
@elpaso , can you have a look at this? Turns out problem was due to broken state of zip items unable to use ogr provider.

File: src/core/qgsdataitem.cpp [hunk 1]
```diff
@@ -809,8 +809,7 @@ QVector<QgsDataItem *> QgsDirectoryItem::createChildren()
       continue;
     }
 
-    // vsizip support was added to GDAL/OGR 1.6 but GDAL_VERSION_NUM not available here
-    //   so we assume it's available anyway
+    if ( fileInfo.suffix() == QLatin1String( "zip" ) )
```
Comment (elpaso) [inline src/core/qgsdataitem.cpp]: Wondering if better to use .compare Qt::CaseInsensitive

File: src/core/qgsdataitem.cpp [hunk 2]
```diff
@@ -1287,51 +1285,8 @@ void QgsZipItem::init()
...
-        // mLibraries.append( library );
-        sDataItemPtr.append( dataItem );
-        sProviderNames.append( k );
-      }
-      else
-      {
-        //QgsDebugMsg ( "Cannot get provider " + k );
-      }
-    }
+    sProviderNames << QStringLiteral( "OGR" ) << QStringLiteral( "OGR" );
```
Comment (elpaso) [inline src/core/qgsdataitem.cpp]: Why adding OGR twice? Did you mean GDAL?

File: src/core/qgsdataitem.cpp [hunk 3]
```diff
@@ -1354,18 +1309,22 @@ QVector<QgsDataItem *> QgsZipItem::createChildren()
...
   // loop over files inside zip
   Q_FOREACH ( const QString &fileName, mZipFileList )
   {
     QFileInfo info( fileName );
     tmpPath = mVsiPrefix + mFilePath + '/' + fileName;
     QgsDebugMsgLevel( "tmpPath = " + tmpPath, 3 );
 
-    // Q_FOREACH( dataItem_t *dataItem, mDataItemPtr )
-    for ( int i = 0; i < sProviderNames.size(); i++ )
+    for ( QgsDataItemProvider *provider : providers )
```
Comment (elpaso) [inline src/core/qgsdataitem.cpp]: const

File: src/providers/gdal/qgsgdaldataitems.h [hunk 4]
```diff
@@ -37,4 +39,15 @@ class QgsGdalLayerItem : public QgsLayerItem
     QString layerName() const override;
 };
 
+//! Provider for GDAL root data item
+class QgsGdalDataItemProvider : public QgsDataItemProvider
+{
+  public:
+    QString name() override { return QStringLiteral( "GDAL" ); }
+
+    int capabilities() override { return QgsDataProvider::File | QgsDataProvider::Dir; }
```
Comment (elpaso) [inline src/providers/gdal/qgsgdaldataitems.h]: Did you omit QgsDataProvider::Net intentionally? (I think it was there in the previous implementation)

**Comment by elpaso** [comment]
LGTM (just left some comments), I didn't test it though.

File: src/core/qgsdataitem.cpp (hunk 2 above)
Comment (nirvn) [reply src/core/qgsdataitem.cpp]: Fixed.

File: src/core/qgsdataitem.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/qgsdataitem.cpp]: Can't const'ify this, won't build.

File: src/providers/gdal/qgsgdaldataitems.h (hunk 4 above)
Comment (nirvn) [reply src/providers/gdal/qgsgdaldataitems.h]: Fixed.

File: src/core/qgsdataitem.cpp [hunk 5]
```diff
@@ -809,8 +808,7 @@ QVector<QgsDataItem *> QgsDirectoryItem::createChildren()
       continue;
     }
 
-    // vsizip support was added to GDAL/OGR 1.6 but GDAL_VERSION_NUM not available here
-    //   so we assume it's available anyway
+    if ( fileInfo.suffix() == QLatin1String( "zip" ) || fileInfo.suffix() == QLatin1String( "tar" ) )
```
Comment (nyalldawson) [inline src/core/qgsdataitem.cpp]: .compare( QLatin1String("zip" ), Qt::CaseInsensitive ) == 0 ...

**Comment by nyalldawson** [comment]
Looking good... love the new icon!

**Pull request merged**